[a2av] Add all_to_all_v API accepting exchange plan

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #168047
* #167792
* #167676
* #167144
* #166896
* #166899
* #166898
* #166774
* #163815
* #163814
* #166772
* #165482
* #165101
* __->__ #164181
* #164164

The `all_to_all_v` API accepts an `ExchangePlan`, so that the user can exchange multiple sets of data without paying the meta data exchange cost.
For example:
```
all_to_all_v(topk_idx_in, topk_idx_out, plan)
all_to_all_v(topk_weights_in, topk_weights_out, plan)
all_to_all_v(tokens_in, tokens_out, plan)
```

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

ngimel commented on torch/distributed/_symmetric_memory/__init__.py:
```diff
@@ -1847,7 +1856,33 @@ def make_a2a_exchange_plan(
+    return ExchangePlan(in_splits, src_offsets, out_splits, dst_offsets)
+
+
+def all_to_all_v(
```
should it be added to __all__?

ngimel commented on torch/csrc/distributed/c10d/symm_mem/nvshmem_extension.cu:
```diff
@@ -945,6 +954,66 @@ void _make_a2a_exchange_plan(
       stream);
 }
 
+void _all_to_all_get(
```
do you need a separate api for this, or just marking some inputs `optional` would allow you to reuse one of the existing all_to_alls?

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

ngimel commented on torch/csrc/distributed/c10d/symm_mem/nvshmem_extension.cu:
```diff
@@ -304,6 +310,68 @@ static int get_a2a_nblocks(size_t size, int world_size, bool intra_node) {
+      &dst_offsets_ptr,
+      &stride_bytes,
+      &team_to_use};
+  C10_CUDA_CHECK(cudaLaunchKernel(
```
ooc why cudaLaunchKernel here and not <<<>>> syntax?

kwen2501 committed:
Update on "[a2av] Add all_to_all_v API accepting exchange plan"


The `all_to_all_v` API accepts an `ExchangePlan`, so that the user can exchange multiple sets of data without paying the meta data exchange cost.
For example:
```
all_to_all_v(topk_idx_in, topk_idx_out, plan)
all_to_all_v(topk_weights_in, topk_weights_out, plan)
all_to_all_v(tokens_in, tokens_out, plan)
```

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]
